File: ROS/src/control/pathplanning/src/boundary_estimation.cpp
#include "boundary_estimation.hpp"

namespace pathplanning {

BoundaryEstimation::BoundaryEstimation(ros::NodeHandle &n)
    : n_(n), frametf_(n), color_connector_(n) {
  this->boundary_pub_ =
      n_.advertise<ugr_msgs::Boundaries>("/output/boundaries", 10);
  this->left_boundary_pub_ =
      n_.advertise<nav_msgs::Path>("/output/debug/left_boundary", 10);
  this->right_boundary_pub_ =
      n_.advertise<nav_msgs::Path>("/output/debug/right_boundary", 10);
  this->map_sub_ = n_.subscribe("/input/local_map", 10,
                                &BoundaryEstimation::receive_new_map, this);
  this->diagnostics_pub = std::unique_ptr<node_fixture::DiagnosticPublisher>(
      new node_fixture::DiagnosticPublisher(n, "CTRL BOUNDARY"));
}

void BoundaryEstimation::receive_new_map(
    const ugr_msgs::ObservationWithCovarianceArrayStamped::ConstPtr &track) {
  std::vector<std::vector<double>> cones;
  for (const auto &obs_with_cov : track->observations) {
    std::vector<double> cone;
    cone.push_back(obs_with_cov.observation.location.x);
    cone.push_back(obs_with_cov.observation.location.y);
    cone.push_back(obs_with_cov.observation.observation_class);
    cones.push_back(cone);
  }

  compute(cones, track->header);
}

void BoundaryEstimation::compute(const std::vector<std::vector<double>> &cones,
                                 const std_msgs::Header &header) {
  std::pair<std::vector<Node *>, std::vector<Node *>> boundaries =
      this->color_connector_.get_color_lines(cones, header);

  std::vector<geometry_msgs::PoseStamped> blue_poses;
  for (const auto &node : boundaries.first) {
    geometry_msgs::PoseStamped pose;

    pose.pose.position.x = node->x;
    pose.pose.position.y = node->y;
    pose.pose.position.z = 0.0;

    pose.pose.orientation.x = 0.0;
    pose.pose.orientation.y = 0.0;
    pose.pose.orientation.z = 0.0;
    pose.pose.orientation.w = 1.0;

    pose.header.frame_id = header.frame_id;
    pose.header.stamp = header.stamp;

    blue_poses.push_back(pose);
  }
  std::vector<geometry_msgs::PoseStamped> yellow_poses;
  for (const auto &node : boundaries.second) {
    geometry_msgs::PoseStamped pose;

    pose.pose.position.x = node->x;
    pose.pose.position.y = node->y;
    pose.pose.position.z = 0.0;

    pose.pose.orientation.x = 0.0;
    pose.pose.orientation.y = 0.0;
    pose.pose.orientation.z = 0.0;
    pose.pose.orientation.w = 1.0;

    pose.header.frame_id = header.frame_id;
    pose.header.stamp = header.stamp;

    yellow_poses.push_back(pose);
  }

  nav_msgs::Path left;
  left.header.frame_id = header.frame_id;
  left.poses = blue_poses;
  left.header.stamp = header.stamp;

  nav_msgs::Path right;
  right.header.frame_id = header.frame_id;
  right.poses = yellow_poses;
  right.header.stamp = header.stamp;

  ugr_msgs::Boundaries boundaries_msg;
  boundaries_msg.left_boundary = left;
  boundaries_msg.right_boundary = right;

  this->boundary_pub_.publish(boundaries_msg);
  this->left_boundary_pub_.publish(left);
  this->right_boundary_pub_.publish(right);
}

} // namespace pathplanning

File: ROS/src/control/boundary_estimation/pkgs/BE/boundary_estimation.py
#!/usr/bin/env python3

import numpy as np
from color_connecter import get_color_lines
from geometry_msgs.msg import PoseStamped
from nav_msgs.msg import Path
from node_fixture.managed_node import ManagedNode
from ugr_msgs.msg import Boundaries, ObservationWithCovarianceArrayStamped


class BoundaryEstimation(ManagedNode):
    def __init__(self):
        super().__init__("boundary_estimation")
        self.boundary_pub = super().AddPublisher(
            "/output/boundaries", Boundaries, queue_size=10
        )
        self.left_boundary_pub = super().AddPublisher(
            "/output/debug/left_boundary", Path, queue_size=10
        )
        self.right_boundary_pub = super().AddPublisher(
            "/output/debug/right_boundary", Path, queue_size=10
        )

        self.spin()

    def doActivate(self):
        self.map_sub = super().AddSubscriber(
            "/input/local_map",
            ObservationWithCovarianceArrayStamped,
            self.receive_new_map,
        )

    def receive_new_map(self, map):
        cones = np.zeros((len(map.observations), 3))
        for i, obs in enumerate(map.observations):
            cones[i] = [
                obs.observation.location.x,
                obs.observation.location.y,
                obs.observation.observation_class,
            ]

        self.compute(cones, map.header)

    def compute(self, cones, header):
        left_boundary, right_boundary = get_color_lines(cones, header)

        left_boundary_msg = Path()
        left_boundary_msg.header = header
        for cone in left_boundary:
            pose = PoseStamped()
            pose.pose.position.x = cone.x
            pose.pose.position.y = cone.y
            pose.pose.position.z = 0

            pose.pose.orientation.x = 0
            pose.pose.orientation.y = 0
            pose.pose.orientation.z = 0
            pose.pose.orientation.w = 1

            pose.header = header
            left_boundary_msg.poses.append(pose)

        right_boundary_msg = Path()
        right_boundary_msg.header = header
        for cone in right_boundary:
            pose = PoseStamped()
            pose.pose.position.x = cone.x
            pose.pose.position.y = cone.y
            pose.pose.position.z = 0

            pose.pose.orientation.x = 0
            pose.pose.orientation.y = 0
            pose.pose.orientation.z = 0
            pose.pose.orientation.w = 1

            pose.header = header
            right_boundary_msg.poses.append(pose)

        boundaries = Boundaries()
        boundaries.left_boundary = left_boundary_msg
        boundaries.right_boundary = right_boundary_msg

        self.boundary_pub.publish(boundaries)
        self.left_boundary_pub.publish(left_boundary_msg)
        self.right_boundary_pub.publish(right_boundary_msg)


BoundaryEstimation()
